src/coreclr/tools/cdac-build-tool/ObjectFileScraper.cs

```diff
@@ -0,0 +1,425 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Buffers.Binary;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.DotNet.Diagnostics.DataContract.BuildTool;
+
+public class ObjectFileScraper
+{
+    public readonly ReadOnlyMemory<byte> MagicLE = new byte[8]{0x44, 0x41, 0x43, 0x42, 0x4C, 0x4F, 0x42, 0x00}; // "DACBLOB\0"
+    public readonly ReadOnlyMemory<byte> MagicBE = new byte[8]{0x00, 0x42, 0x4F, 0x4C, 0x42, 0x43, 0x41, 0x44};
+
+    private readonly DataDescriptorModel.Builder _builder;
+
+    public bool Verbose {get;}
+    public ObjectFileScraper(bool verbose, DataDescriptorModel.Builder builder)
+    {
+        Verbose = verbose;
+        _builder = builder;
+    }
+
+    public async Task<bool> ScrapeInput(string inputPath, CancellationToken token)
+    {
+        using var file = File.Open(inputPath, FileMode.Open, FileAccess.Read, FileShare.Read);
+        var r = await FindMagic(file, token);
+
+        if (!r.Found) {
+            return false;
+        }
+        if (Verbose) {
+            Console.WriteLine ($"{inputPath}: magic at {r.Position}");
+        }
+        file.Seek(r.Position + (long)MagicLE.Length, SeekOrigin.Begin); // skip magic
+        var headerStart = file.Position;
+        var header = ReadHeader(file, r.LittleEndian);
+        if (Verbose) {
+            Console.WriteLine ($"{inputPath}: namesStart = {header.NamesStart}");
+            Console.WriteLine ($"{inputPath}: typeCount = {header.TypeCount}");
+            Console.WriteLine ($"{inputPath}: fieldPoolCount = {header.FieldPoolCount}");
+        }
+        var content = ReadContent(file, headerStart, header, r.LittleEndian);
+        content.AddToModel(_builder);
+        return true;
+    }
+
+    struct MagicResult {
+        public bool Found {get; init;}
+        public long Position {get; init;}
+        public bool LittleEndian {get; init;}
+    }
+
+    private async Task<MagicResult> FindMagic(Stream stream, CancellationToken token)
+    {
+        var buf = new byte[4096];
+        long pos = stream.Position;
+        while (true) {
+            token.ThrowIfCancellationRequested();
+            int bytesRead = await stream.ReadAsync(buf, 0, buf.Length, token);
+            if (bytesRead == 0)
+                return new (){Found = false, Position = 0, LittleEndian = true};
+            // FIXME: what if magic spans a buffer boundary
```

Comment:
This sort of tool is meant to process a single object file artifact. As such I think it makes sense to just read in the entire binary blob and entirely avoid these sorts of issues. I can't imagine any needed object file artifact being more than 1 MB. Even if it becomes 100 MB, we can handle that.